cpp/include/Ice/AsyncResponseHandler.h

```diff
@@ -0,0 +1,87 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+#ifndef ICE_ASYNC_RESPONSE_HANDLER_H
+#define ICE_ASYNC_RESPONSE_HANDLER_H
+
+#include "Current.h"
+#include "OutgoingResponse.h"
+#include "LocalException.h"
+
+#include <atomic>
+
+namespace IceInternal
+{
+    // This class helps with the implementation of the AMD response and exception callbacks. It allows the dispatch
+    // thread and these two callbacks to share the same sendResponse and Current objects, and ensures sendResponse is
+    // called exactly once.
+    class AsyncResponseHandler final
+    {
+    public:
+        // This class typically holds a _copy_ of the incoming request current object.
+        AsyncResponseHandler(std::function<void(Ice::OutgoingResponse)> sendResponse, Ice::Current current)
```

Comment:
I probably miss something but I don't clearly understand why it's useful. Is it only about ensuring that we don't send twice the response? Could we have a internal flag on `OutgoingResponse` for this purpose instead?

Reply:
> Could we have a internal flag on OutgoingResponse for this purpose instead?

How would this work?

OutgoingResponse is a value that each thread can construct independently. Your response callback and exception callback could each construct an OutgoingResponse and pass it to sendResponse.